rskj-core/src/main/java/co/rsk/net/NodeBlockProcessor.java

```diff
@@ -148,208 +121,219 @@ private boolean hasHeader(@Nonnull final byte[] hash) {
-        if (sender == null)
-            return;
-
-        if (unknownBlockHashes.containsKey(hash)) {
-            int counter = unknownBlockHashes.get(hash).intValue();
-
-            counter++;
+    /**
+     * processBodyRequest sends the requested block body to a peer if it is available.
+     *
+     * @param sender the sender of the BodyRequest message.
+     * @param requestId the id of the request
+     * @param hash   the requested block's hash.
+     */
+    @Override
+    public void processBodyRequest(@Nonnull final MessageChannel sender, long requestId, @Nonnull final byte[] hash) {
+        logger.trace("Processing body request {} {} from {}", requestId, Hex.toHexString(hash).substring(0, 10), sender.getPeerNodeID().toString());
+        final Block block = blockSyncService.getBlockFromStoreOrBlockchain(hash);
 
-            if (counter <= 20) {
-                unknownBlockHashes.put(hash, new Integer(counter));
-                return;
-            }
+        if (block == null) {
+            // Don't waste time sending an empty response.
+            return;
         }
 
-        unknownBlockHashes.put(hash, new Integer(1));
-
-        logger.trace("Missing block " + hash.toString().substring(0, 10));
-
-        sender.sendMessage(new GetBlockMessage(hash.getData()));
-
-        return;
+        Message responseMessage = new BodyResponseMessage(requestId, block.getTransactionsList(), block.getUncleList());
+        sender.sendMessage(responseMessage);
     }
 
     /**
-     * processStatus processes a Status containing another node's status (its bestBlock).
-     * If the sender has a better best block, it will be requested.
-     * Otherwise, all the blocks that the sender is missing will be sent to it.
-     *
-     * @param sender the message sender. This should be the node that sent the status message.
-     * @param status The status message containing the other node's best block.
+     * processBlockHashRequest sends the requested block body to a peer if it is available.
+     *  @param sender the sender of the BlockHashRequest message.
+     * @param requestId the id of the request
+     * @param height   the requested block's hash.
      */
     @Override
-    public void processStatus(@Nonnull final MessageSender sender, @Nonnull final Status status) {
-        logger.trace("Processing status " + status.getBestBlockNumber() + " " + Hex.toHexString(status.getBestBlockHash()).substring(0, 10) + " from " + sender.getNodeID().toString());
-
-        final byte[] hash = status.getBestBlockHash();
-        nodeInformation.addBlockToNode(new ByteArrayWrapper(hash), sender.getNodeID());
+    public void processBlockHashRequest(@Nonnull final MessageChannel sender, long requestId, long height) {
+        logger.trace("Processing block hash request {} {} from {}", requestId, height, sender.getPeerNodeID().toString());
+        final Block block = this.getBlockFromBlockchainStore(height);
```

Comment:
We can add a height > 0 here

Reply:
Added guard to avoid retrieving genesis block :+1: